### PR TITLE
Dataflow flex-templates resource

### PR DESCRIPTION
**Pull request by geojaz**
**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_dataflow_flex_template_job`
```

We'd like to get the beta provider for terraform to support [Dataflow Flex-Templates](https://cloud.google.com/dataflow/docs/guides/templates/using-flex-templates). This is based on the hand-written `dataflow_job` resource and appears to work for flex-template CRUD activities.

I will continue to iterate on this next week with some tests and to finish off the docs. I'll ping whn it's good to be reviewed.

FYI @pupamanyu @WillBeebe @cxhercules

[drive-by edit by danawillow:] Fixes https://github.com/terraform-providers/terraform-provider-google/issues/6656

### CODE REVIEW

**Comment by geojaz** [comment]
Upon further thinking, I believe it makes most sense to combine `resource_dataflow_job` and `resource_dataflow_flex_template_job`, since they hit the same GCP endpoint (with different params).... I'll refactor the new logic to pull it into `resource_dataflow_job'.

**Comment by geojaz** [comment]
> Do you mind showing me how Flex Templates end up using the same endpoint? It seems they've got a different gcloud resource (which are normally strongly linked to endpoints) as well as [this CURL example](https://cloud.google.com/dataflow/docs/guides/templates/using-flex-templates#running_a_flex_template_pipeline) which appears to hit a flex-specific endpoint: `https://dataflow.googleapis.com/v1b3/projects/$PROJECT/locations/$LOCATION/flexTemplates`

**Comment by geojaz** [comment]
@rileykarson You're right, I was having a brain fart. I've retracted my proposal to combine these resources and will finish off `resource_dataflex_flex_template_job`.

**Comment by geojaz** [comment]
@rileykarson @emilymye I'd love to get some feedback on the PR now. I can see that I have a failure in the cloudbuild trigger, but I'm unable to see what it is.

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+				Computed: true,
+			},
+
+			"labels": {
+				Type:             schema.TypeMap,
+				Optional:         true,
+				DiffSuppressFunc: resourceDataflowJobLabelDiffSuppress,
+			},
+
+			"name": {
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:44]: nit: mind ordering fields Required -> Optional (and Optional+Computed) -> Computed?

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		return updateErr
+	}, time.Minute*time.Duration(5), isDataflowJobUpdateRetryableError)
+	if err != nil {
+		return err
+	}
+	d.SetId(job.Id)
+	return resourceDataflowJobRead(d, meta)
+}
+
+// resourceDataflowFlexTemplateJobLaunchTemplate Launches a Flex Template Job.
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:168]: I'd slightly prefer that we inline both methods below. While it's not DRY, it matches the convention in the rest of the provider. If you feel strongly otherwise, feel free to ignore this suggestion.

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		return err
+	}
+
+	request := buildLaunchFlexTemplateRequest(d, config)
+
+	job, err := resourceDataflowFlexTemplateJobLaunchTemplate(config, project, region, &request)
+	if err != nil {
+		return err
+	}
+	d.SetId(job.Id)
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:99]: Can you also store this value in the job_id computed field? Here & other SetId occurences

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	if err != nil {
+		return err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:118]: nit: jobID is a little more clear of a variable name

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+	var job *dataflow.Job
+	err = retryTimeDuration(func() (updateErr error) {
+		job, updateErr = resourceDataflowFlexTemplateJobLaunchTemplate(config, project, region, &request)
+		return updateErr
+	}, time.Minute*time.Duration(5), isDataflowJobUpdateRetryableError)
+	if err != nil {
+		return err
+	}
+	d.SetId(job.Id)
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:164]: just curious- is the id of the job expected to change here?

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 6]
```diff
@@ -0,0 +1,187 @@
+<% autogen_exception -%>
+<% unless version == 'ga' -%>
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:2]: You'll want to move the `package google` bit up above the version check, Cloud Build is complaining because Go fails to compile packageless files (which the `google` provider one is right now)

File: third_party/terraform/tests/resource_dataflow_flex_template_job_test.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
+package google
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_dataflow_flex_template_job_test.go:1]: This file will also need versions guards

File: third_party/terraform/utils/provider.go.erb [hunk 8]
```diff
@@ -317,6 +317,7 @@ end     # products.each do
 				"google_container_node_pool":                   resourceContainerNodePool(),
 				"google_container_registry":                    resourceContainerRegistry(),
 				"google_dataflow_job":                          resourceDataflowJob(),
+				"google_dataflow_flex_template_job":            resourceDataflowFlexTemplateJob(),
```
Comment (rileykarson) [inline third_party/terraform/utils/provider.go.erb:320]: & version guard this entry in the map. There should be similar examples above/below.

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb (hunk 6 above)
Comment (geojaz) [reply third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:2]: ahh! that's how it works :)

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	}
+	response, err := config.clientDataflow.Projects.Locations.FlexTemplates.Launch(project, region, &request).Do()
+	if err != nil {
+		return err
+	}
+	job := response.Job
+	d.SetId(job.Id)
+	d.Set("job_id", job.Id)
+
+	return resourceDataflowFlexTemplateJobRead(d, meta)
```
Comment (jaketf) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:105]: If a job immediately fails (bad runtime arg, permissions issue) what happens? 
will this resource silently be dropped in [L136-137](https://github.com/GoogleCloudPlatform/magic-modules/pull/3520/files#diff-7411ff0ee82c3c225d802f7f82acdfb1R136-R137)?
Would it be better to block until pipeline enters running state in the create?

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+		d.SetId("")
+		return nil
+	}
+	d.SetId(job.Id)
+	d.Set("job_id", job.Id)
+
+	return nil
+}
+
+// resourceDataflowFlexTemplateJobUpdateByReplacement updates a Flex Template job by replacement.
```
Comment (jaketf) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:145]: this should have TODO to be replaced when flex templates support update.

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	request := dataflow.LaunchFlexTemplateRequest{
+		LaunchParameter: &dataflow.LaunchFlexTemplateParameter{
+			ContainerSpecGcsPath: d.Get("container_spec_gcs_path").(string),
+			JobName:              d.Get("name").(string),
+			Parameters:           expandStringMap(d, "parameters"),
+		},
+	}
+	var job *dataflow.Job
+	err = retryTimeDuration(func() (updateErr error) {
+		response, err := config.clientDataflow.Projects.Locations.FlexTemplates.Launch(project, region, &request).Do()
```
Comment (jaketf) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:168]: This seems like it should stop the old job respecting behavior specified in `on_delete`.
Is this method invoked only after destroy in an update scenario?

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb (hunk 9 above)
Comment (rileykarson) [reply third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:105]: After initial creation, if the resource reaches an error state before that check, if will register as having failed to create. If it reaches the error state afterwards, it will be removed from Terraform state on a subsequent plan, and Terraform will indicate that it needs to be recreated- not exactly silent, but not exactly noisy either.

Stateful resource management can be done additively, so I'd prefer it doesn't get added in the initial PR. A followup is fine, although completely optional- https://github.com/terraform-providers/terraform-provider-google-beta/blob/master/google-beta/resource_container_cluster.go#L2179-L2203 is the most effective approach I've found for working with states.

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+		return nil
+	}
+	d.SetId(job.Id)
+	d.Set("job_id", job.Id)
+
+	return nil
+}
+
+// resourceDataflowFlexTemplateJobUpdateByReplacement updates a Flex Template job by replacement.
+func resourceDataflowFlexTemplateJobUpdateByReplacement(d *schema.ResourceData, meta interface{}) error {
```
[the remark this answers is not in the document]
Comment (rileykarson) [reply third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:146]: Is there documentation available on what pipeline update will look like? Changing the update method under-the-hood could count as a breaking change, so it may be better to release without update and add that once it's available (which would just be an enhancement).

Otherwise, issues > TODOs 🙂

[the remark this answers is not in the document]
Comment (c2thorn) [reply third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:146]: +1 to releasing without an update. Before adding Terraform support for update-by-replacement, the standard Dataflow job implementation had no update method and used `ForceNew` on every field.

File: third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+resource "google_dataflow_flex_template_job" "big_data" {
+  name = "%s"
+  container_spec_gcs_path = "%s"
+  on_delete = "cancel"
+}
+`, bucket, bucket, flexTemplateContent(), bucket, job)
+}
+
+func flexTemplateContent() string {
+	return `
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb:56]: To add a multiline string, you'll want to use heredoc syntax: https://www.terraform.io/docs/configuration/expressions.html#string-literals

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb (hunk 11 above)
Comment (geojaz) [reply third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb]: We can address this in the next iteration... for now I've removed the update logic and added some comments to clarify the behaviors.

**Comment by aaltay** [comment]
Thank you for adding me. Do you want me to review something specific?

Adding, people who directly work on templates: @azurezyq @arvindram03

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+			"on_delete": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"cancel", "drain"}, false),
+				Optional:     true,
+				Default:      "drain",
+			},
+
+			"labels": {
+				Type:             schema.TypeMap,
+				Optional:         true,
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:51]: Can you mark this as `ForceNew: true,`?

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+// NOTE: resource_dataflow_flex_template currently does not support updating existing jobs.
+// Changing any non-computed field will result in the job being deleted (according to its
+// on_delete policy) and recreated with the updated parameters.
+
+// resourceDataflowFlexTemplateJob defines the schema for Dataflow FlexTemplate jobs.
+func resourceDataflowFlexTemplateJob() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDataflowFlexTemplateJobCreate,
+		Read:   resourceDataflowFlexTemplateJobRead,
+		// Update: resourceDataflowFlexTemplateJobUpdateByReplacement,
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:24]: This comment can be deleted

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+				ForceNew: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"on_delete": {
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:42]: I _believe_ we want a blank update method defined. That will make Terraform persist the value for `on_delete` into state. If there's no method at all, it requires this field to recreate the resource.

File: third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+func testAccDataflowFlowFlexTemplateJob_basic(bucket, job string) string {
+	return fmt.Sprintf(`
+resource "google_storage_bucket" "temp" {
+  name = "%s"
+  force_destroy = true
+}
+
+resource "google_storage_bucket_object" "flex_template" {
+	name   = "flex_template.json"
+	bucket = "%s"
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb:43]:
```suggestion
	bucket = google_storage_bucket.temp.name
```

File: third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+	content = "%s"
+}
+
+resource "google_dataflow_flex_template_job" "big_data" {
+  name = "%s"
+  container_spec_gcs_path = "%s"
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb:49]:
```suggestion
  container_spec_gcs_path = "${google_storage_bucket.temp.url}/${google_storage_bucket_object.flex_template.name}"
```

File: third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+  on_delete = "cancel"
+}
+`, bucket, bucket, flexTemplateContent(), bucket, job)
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb:52]:
```suggestion
`, bucket, job)
```

File: third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func flexTemplateContent() string {
+	return `
+<<EOF
+{
+	"name": "Streaming Beam SQL",
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_dataflow_flex_template_job_test.go.erb:59]: Is this a complete flex template? I get the following error after making the changes noted above:  `Error: googleapi: Error 400: (165707124cd2c54b): Docker image not specified in the template file., badRequest`

File: third_party/terraform/website/docs/r/dataflow_flex_template_job.html.markdown [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+There are many types of Dataflow jobs.  Some Dataflow jobs run constantly, getting new data from (e.g.) a GCS bucket, and outputting data continuously.  Some jobs process a set amount of data then terminate.  All jobs can fail while running due to programming errors or other issues.  In this way, Dataflow jobs are different from most other Terraform / Google resources.
+
+The Dataflow resource is considered 'existing' while it is in a nonterminal state.  If it reaches a terminal state (e.g. 'FAILED', 'COMPLETE', 'CANCELLED'), it will be recreated on the next 'apply'.  This is as expected for jobs which run continuously, but may surprise users who use this resource for other kinds of Dataflow jobs.
+
+A Dataflow job which is 'destroyed' may be "cancelled" or "drained".  If "cancelled", the job terminates - any data written remains where it is, but no new data will be processed.  If "drained", no new data will enter the pipeline, but any data currently in the pipeline will finish being processed.  The default is "cancelled", but if a user sets `on_delete` to `"drain"` in the configuration, you may experience a long wait for your `terraform destroy` to complete.
```
Comment (aaltay) [inline third_party/terraform/website/docs/r/dataflow_flex_template_job.html.markdown:35]: You could link to : https://cloud.google.com/dataflow/docs/guides/stopping-a-pipeline for effects of cancel vs drain.

I do not know if `destroyed` is TF terminology. Dataflow term for this is `stopping`.

File: third_party/terraform/website/docs/r/dataflow_flex_template_job.html.markdown [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+  container_spec_gcs_path = "gs://my-bucket/templates/template.json"
+  parameters = {
+    inputSubscription = "messages"
+  }
+}
+```
+
+[ To Come ...]
+## Note on "destroy" / "apply"
+There are many types of Dataflow jobs.  Some Dataflow jobs run constantly, getting new data from (e.g.) a GCS bucket, and outputting data continuously.  Some jobs process a set amount of data then terminate.  All jobs can fail while running due to programming errors or other issues.  In this way, Dataflow jobs are different from most other Terraform / Google resources.
```
Comment (aaltay) [inline third_party/terraform/website/docs/r/dataflow_flex_template_job.html.markdown:31]: Dataflow terminology for these two types of jobs are `streaming` and `batch` jobs respectively.

File: third_party/terraform/website/docs/r/dataflow_flex_template_job.html.markdown [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+The Dataflow resource is considered 'existing' while it is in a nonterminal state.  If it reaches a terminal state (e.g. 'FAILED', 'COMPLETE', 'CANCELLED'), it will be recreated on the next 'apply'.  This is as expected for jobs which run continuously, but may surprise users who use this resource for other kinds of Dataflow jobs.
```
Comment (aaltay) [inline third_party/terraform/website/docs/r/dataflow_flex_template_job.html.markdown:33]: Suggestion: Replace `for other kinds of Dataflow jobs.` with `for Dataflow batch jobs.`

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"cancel", "drain"}, false),
+				Optional:     true,
+				Default:      "drain",
```
Comment (aaltay) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:44]: Doc below says the default is cancel. (`The default is "cancelled",`)

File: third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+	d.Set("name", job.Name)
+	d.Set("project", project)
+	d.Set("labels", job.Labels)
+
+	if _, ok := dataflowTerminalStatesMap[job.CurrentState]; ok {
+		log.Printf("[DEBUG] Removing resource '%s' because it is in state %s.\n", job.Name, job.CurrentState)
+		d.SetId("")
+		return nil
+	}
+	d.SetId(job.Id)
```
Comment (aaltay) [inline third_party/terraform/resources/resource_dataflow_flex_template_job.go.erb:144]: Why is this line needed L127, read this info from this field already?

**Comment by rileykarson** [comment]
I copied the changes from this PR into https://github.com/GoogleCloudPlatform/magic-modules/pull/3772, superseding this one. @geojaz do you mind responding to the [googlebot comment](https://github.com/GoogleCloudPlatform/magic-modules/pull/3772#issuecomment-662618659) when you have the chance there? I squashed your commits into a single one, for ease of reviewability.
